stream-helpers/README.md

```diff
@@ -0,0 +1,227 @@
+
+  for (const batch of yield* each(stream)) {
+    console.log(batch); // Items received within 1 second
+    yield* each.next();
+  }
+});
+
+// Example: Combined batching
+function* exampleCombined(stream: Stream<number, unknown>) {
+
+  const batched = batch({
+    maxSize: 5,
+    maxTime: 1000,
+  });
+
+  for (const batch of yield* each(batched(stream))) {
+    console.log(batch); // Up to 5 items within 1 second
+    yield* each.next();
+  }
+});
+```
+
+### Valve
+
+Allows to apply backpressure to the source stream to prevent overwhelming the
+downstream consumer. This is useful with any stream that generates items faster
+than the consumer can consume them. It was originally designed for use with
+Kafka where the producer can cause the service to run out of memory when the
+producer produces many faster than the consumer to process the messages. It can
+be used as a buffer for any infinite stream.
+
+```typescript
+import { valve } from "@effectionx/stream-helpers";
+import { each } from "effection";
+
+function* example() {
+  const regulated = valve({
+    // buffer size threshold when close operation will invoked
+    closeAt: 1000,
+    *close() {
+      // pause the source stream
+    },
+
+    // buffer size threshold when open operation will be invoked
+    openAt: 100,
+    *open() {
+      // resume the source stream
+    },
+  });
+
+  for (const value of yield* each(regulated(stream))) {
+    console.log(value);
+    yield* each.next();
+  }
+}
+```
+
+### Composing stream helpers
+
+You can use a simple `pipe()` to compose a series of stream helpers together. In
+this example, we use one from [remeda](https://remedajs.com/docs/#pipe),
+
+```typescript
+import { batch, filter, map, valve } from "@effectionx/stream-helpers";
+import { each } from "effection";
+// any standard pipe function should work
+import { pipe } from "remeda";
+
+function* example(source: Stream<number, unknown>) {
+  // Compose stream helpers using pipe
+  const stream = pipe(
+    source,
+    valve({ open, close, openAt: 100, closeAt: 100 }),
+    filter(function* (x) {
+      return x > 0;
+    }),
+    map(function* (x) {
+      return x * 20;
+    }),
+    batch({ maxSize: 50 }),
+  );
+
+  for (const value of yield* each(stream)) {
+    console.log(value);
+    yield* each.next();
+  }
+}
+```
+
+## Testing Streams
+
+The library includes testing utilities to help you test your stream processing
+code. These are available in `@effectionx/stream-helpers/test-helpers` export.
+
+### Faucet
+
+The `createFaucet` function creates a stream that can be used to test the
+behavior of streams that use backpressure. It's particularly useful in tests
+where you need a controllable source stream.
+
+```typescript
+import { createFaucet } from "@effectionx/stream-helpers/test-helpers";
+import { each, run, spawn } from "effection";
+
+await run(function* () {
+  const faucet = yield* createFaucet<number>({ open: true });
```

Comment:
typically `createX()` implies that `X` is stateless, whereas `useX()` implies that this is a resource.